### ISSUE TITLE
PassableTech in Features and Terrains

### ISSUE THREAD

**Issue by JFDaly** [LoneGazebo/Community-Patch-DLL#693]
Hopefully another small and easy request for a 'PassableTech' tag to be added to the Features and Terrains tag, in order to negate the 'Impassable' tag in those tables. In my mind, this would be used to handle my Jungle terrain in ExCE, which is impassable until Optics. Thanks.

**Comment by LoneGazebo** [comment]
I'll take a look, but probably not for this upcoming version. I know ilteroi did a lot with impassable elements in the previous versions, and may have condensed them in a way that makes negation tricky. I'll report back.

**Comment by ilteroi** [comment]
well, all i can say is it's tricky.

i aborted my previous attempt to handle impassability in a logical manner and concluded that this part of the code should better not be touched. look for MOD_BALANCE_CORE_SANE_IMPASSABILITY ifdefs :)

**Comment by JFDaly** [comment]
Ah, well, if that's that tricky, then by no means worry about it. I can make do in Lua.

**Comment by LoneGazebo** [comment]
I'm tinkering with this.

**Comment by LoneGazebo** [comment]
Alright, I think I've got something. There'll be two new columns for terrains and features:

-- Allows you to set a tech that makes an impassable terrain/feature element passable.
ALTER TABLE Features ADD COLUMN 'PassableTechFeature' TEXT DEFAULT NULL;

-- Allows you to set a tech that makes an impassable terrain/feature element passable.
ALTER TABLE Terrains ADD COLUMN 'PassableTechTerrain' TEXT DEFAULT NULL;

Also, in the isImpassable lua function, you can now input a team value to get whether or not the plot is impassable for you (using the BARBARIAN_TEAM value will get return true for all plots except mountains and ice).

I'm still testing, of course, but the good news is that this will actually let the AI know how your functionality works (formerly, I imagine, any attempts to spoof impassability were actually breaking the AI's pathfinder from time to time).

G

**Comment by LoneGazebo** [comment]
Not sure if you want to, JFD, but you could edit the techbuttoninclude.lua and civilopedia.lua to generate icons/text in the tech tree/civilopedia for techs that make terrains/features passable. If you do this,mind dropping me the text so I can add it to the CP's versions of these files?

**Comment by JFDaly** [comment]
Sure thing. I've already set up the tech button to do that for Jungle (I've actually set up a table that'll allow those icons to be added without hardcoding them in the tech button). I'll see about doing it for the civilopedia, too - the included civilopedia is a bit outdated, and needs to be replaced with another that I created to support my alternative leaders civs better. But I'll definitely post them to you and you can include them in the CP.

Thanks! This looks promising.

**Comment by JFDaly** [comment]
Here is the [TechButtonIncludes replacement](https://www.dropbox.com/s/saj7qfaav3svh1s/TechButtonInclude.rar?dl=0). Just import all three Lua files into VFS and add the SQL to the database.

In the SQL, I've already set up some of the ones specific to CBP. Follow that format if you want to add any others. EUI is automatically supported - the UnitActions column defines a mission from which the tech button will pull the icon, if EUI is active.

**Comment by LoneGazebo** [comment]
Alright, I've got a test version up for you.

To get this to work, you need to set terrains/features impassable, and then set the new value for them for a tech unlock. Like so:

-- Test
UPDATE Terrains
SET Impassable = 'true'
WHERE Type = 'TERRAIN_DESERT';

UPDATE Terrains
SET PassableTechTerrain = 'TECH_POTTERY'
WHERE Type = 'TERRAIN_DESERT';

UPDATE Features
SET Impassable = 'true'
WHERE Type = 'FEATURE_JUNGLE';

UPDATE Features
SET PassableTechFeature = 'TECH_THE_WHEEL'
WHERE Type = 'FEATURE_JUNGLE';

To give units the ability to override impassable elements before you have the tech, you need to use these two tables:

<Table name="UnitPromotions_Terrains">
		<Column name="PromotionType" type="text" reference="UnitPromotions(Type)"/>
		<Column name="PassableTech" type="text" reference="Technologies(Type)"/>
	</Table>
	<Table name="UnitPromotions_Features">
		<Column name="PassableTech" type="text" reference="Technologies(Type)"/>
		<Column name="PediaType" type="text"/>
	</Table>

All that matters is that you set the passable tech (probably to agriculture) and then the terrain/feature you want to override. I think you already use these tables, so should be good to go. The spawnadjacentfreepromotion stuff will hook into this, so if you give a jungle that element, and then make it give a promotion that has the passabletech stuff above, it'll allow you to (theoretically) avoid any lua for impassibility.

This should, theoretically, work. Theoretically. I need you to test it out, make sure that the permutations are operating as desired.

Here's the link: https://mega.nz/#!SI0zCTLD!QVH3lM1DoX740L06qmUdqBHITNR49osJiT2BoNM0nNQ


So when should I expect JFD's Gazebo Civilization? :)

**Comment by JFDaly** [comment]
Awesome. It almost all works as far as I tested - Jungle was impassable until Optics, but for units with the right promotion, it was passable since Agriculture. Also, strictly for testing of the terrains part, snow was impassable until Chivalry, and again the right promotion allowed a unit to bypass this.

Well, I have always intended to include you as a new civ in the World According to Modders scenario (a mod I made last year for April Fools, with the major civ modders+Whoward as civs), once I get around to updating it.
